agp-handlers/agp-handler-api/src/main/kotlin/slack/gradle/agp/AgpHandlerFactory.kt

```diff
@@ -24,7 +26,42 @@ package slack.gradle.agp
  */
 public interface AgpHandlerFactory {
   public val minVersion: VersionNumber
-  /** Attempts to get the current AGP version or throws and exception if it cannot. */
+  /** Attempts to ascertain the current AGP version or throws and exception if it cannot. */
   public fun currentVersion(): String
-  public fun create(): AgpHandler
+  /**
+   * Creates a new [AgpHandler] instance for the current AGP version if [currentVersion] was deemed
+   * to match.
+   */
+  public fun createHandler(): AgpHandler
+  /**
+   * Creates a new [AgpSettingsHandler] instance for the current AGP version if [currentVersion] was
+   * deemed to match.
+   */
+  public fun createSettingsHandler(): AgpSettingsHandler = AgpSettingsHandler.NoOp
+
+  public companion object {
+    public fun load(): AgpHandlerFactory {
+      /**
+       * Load handlers and pick the highest compatible version (by [AgpHandlerFactory.minVersion])
+       */
+      return ServiceLoader.load(AgpHandlerFactory::class.java)
+        .iterator()
+        .asSequence()
+        .mapNotNull { factory ->
+          // Filter out any factories that can't compute the AGP version, as
+          // they're _definitely_ not compatible
+          try {
+            FactoryData(VersionNumber.parse(factory.currentVersion()), factory)
+          } catch (t: Throwable) {
+            null
+          }
+        }
+        .filter { (agpVersion, factory) -> agpVersion.baseVersion >= factory.minVersion }
+        .maxByOrNull { (_, factory) -> factory.minVersion }
+        ?.factory
+        ?: error("Unrecognized AGP version!")
```

Comment:
Is this good enough, or can the error message contain a more actionable message?

Reply:
Good enough for now, it's what we had before too